crates/isograph_compiler/src/source_files.rs

```diff
@@ -0,0 +1,353 @@
+                }
+                SourceEventKind::Remove(path) => {
+                    if let Err(e) = self.handle_remove(path, change_kind) {
+                        errors.push(e);
+                    }
+                }
+            }
+        }
+
+        if !errors.is_empty() {
+            Err(BatchCompileError::MultipleErrors { messages: errors })
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn handle_create_or_modify(
+        &mut self,
+        config: &CompilerConfig,
+        path: &PathBuf,
+        change_kind: &ChangedFileKind,
+    ) -> Result<(), BatchCompileError> {
+        match change_kind {
+            ChangedFileKind::Schema => {
+                self.schema = read_and_parse_graphql_schema(&config.schema)?;
+            }
+            ChangedFileKind::SchemaExtension => {
+                let (file_path, document) = read_and_parse_schema_extensions(path)?;
+                match self.schema_extensions.entry(file_path) {
+                    Entry::Occupied(occupied_entry) => {
+                        let value = occupied_entry.into_mut();
+                        *value = document;
+                    }
+                    Entry::Vacant(vacant_entry) => {
+                        vacant_entry.insert(document);
+                    }
+                };
+            }
+            ChangedFileKind::JavaScriptSourceFile => {
+                let canonicalized_root_path = get_canonicalized_root_path(&config.project_root)?;
+                let (path_buf, file_content) =
+                    read_file(path.to_path_buf(), &canonicalized_root_path)?;
+                let (file_path, iso_literals) =
+                    read_and_parse_iso_literals(path_buf, file_content, &canonicalized_root_path)?;
+                if iso_literals.is_empty() {
+                    self.contains_iso.remove(&file_path);
+                } else {
+                    match self.contains_iso.entry(file_path) {
+                        Entry::Occupied(occupied_entry) => {
+                            let value = occupied_entry.into_mut();
+                            *value = iso_literals;
+                        }
+                        Entry::Vacant(vacant_entry) => {
+                            vacant_entry.insert(iso_literals);
+                        }
+                    };
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn handle_rename(
+        &mut self,
+        config: &CompilerConfig,
+        source_path: &Path,
+        target_path: &PathBuf,
+        change_kind: &ChangedFileKind,
+    ) -> Result<(), BatchCompileError> {
+        match change_kind {
+            ChangedFileKind::Schema => {
+                if config.schema != *target_path {
+                    return Err(BatchCompileError::SchemaNotFound);
+                }
+            }
+            ChangedFileKind::SchemaExtension => {
+                if config.schema_extensions.contains(target_path) {
```

Comment:
should we handle schema extension renames? At least right now, they all have to be explicitly named in the config... I haven't thought much about this.

Reply:
Yeah, it is safe to rename schema extension if we add watch for config file. In that case an extension stops being watched and will be recreated if user reverts the name or changes the config. And we will see errors if something is used in iso literals. Although it's not consistent with batch mode where we panic if schema extension is absent